virttest.qemu_monitor: Automatic fallback to "x-" mig capability

ldoktor opened the pull request:
In order to support various qemu versions, let's always try using "x-"
or non "x-" variant of the migration capability.

Signed-off-by: Lukáš Doktor <ldoktor@redhat.com>

ldoktor commented:
not yet complete, I need to taclke params as well...

ldoktor commented:
@Hao-Liu @luckyh without this patch we can't test the latest qemu-kvm migrate.multifd. Would you please take a look? I tried it with old and new qemu with qmp and hmp by: `avocado --show all run migrate.default.tcp..with_multifd` (note Fedora 30 qemu fails to migrate with multifd due to known issue, but the caps/params are set properly)

luckyh commented on virttest/qemu_monitor.py:
```diff
@@ -629,6 +631,50 @@ def wait_for_migrate_progress(self, target):
+    def _get_migrate_capability(self, capability):
+        """
+        Verify the $capability is listed in migrate-capabilitys. If not try
+        x-/non-x- version. In case none is supported, return the original param
```
Suggest making the behavior configurable: by enabled, automatically chose supported options; by disabled, uses user specified ones.

ldoktor replied:
Would it be fine to leave it as is and wait for someone actively requiring this?

luckyh replied:
It is acceptable to me that make the `x` to `non-x` conversion without using a param as the flag, but I think it's not good to do so for the counter one because verifying the existence of formal capabilities/parameters is one of the checkpoints, and it should be a positive test.

ldoktor replied:
Do you have an example of such test? I mean do we want it by default to be disabled, or enabled? Do we want the `vm.migrate` to allow enabling/disabling it (therefore yet another argument)?

luckyh replied:
Let me summarize my perspective: let tests do what they exactly want to do, not the modified version provided by the vt framework, unless it is what they required.

So,
> Do you have an example of such test?

Let's take a look at the test which is supposed to be benefited,
```
# qemu/tests/cfg/migrate.cfg
...
                - with_multifd:
                    only tcp,unix,fd
                    migrate_capabilities = "{'x-multifd': 'on'}"
                    migrate_parameters = "{'x-multifd-channels': 8, 'x-multifd-page-count': 64}"
                    target_migrate_parameters = "{'x-multifd-channels': 8, 'x-multifd-page-count': 64}"
...
```
at least I can't say it doesn't care about checking the existence of capabilities/parameters because the author left nothing but codes. So as a result the feature introduced by this PR would redefine the usage of the above three at the bottom, that's ok, but please leave a comment or write a doc to clarify that, will you? not only for this one, but also for the rest.

> I mean do we want it by default to be disabled, or enabled?

To my perspective, I'd suggest making it be disabled as the default, but if you think the majority of tests do require this feature you can make it be enabled. With the param, for people who don't want it, like me, the feature can be disabled easily with using their own customized configurations.

> Do we want the vm.migrate to allow enabling/disabling it (therefore yet another argument)?

Introducing another argument to the `vm.migrate()` is just one of the approaches, you can do it by introducing another attribute to the `VM` class as a flag (e.g. `vm.auto_x_feature_conversion`) and set it according to user's configuration in `__init__()`, and then reference it in `migrate()`. Please be noted that the `x-` things handled in this PR are not only the ones to be benefited, there are other experimental features, for example the things are handled in https://github.com/avocado-framework/avocado-vt/pull/1932, so just don't worry about it is an "unnecessary" flag.

ldoktor replied:
Well, I guess this feature is closer to selecting the right machine type, cpu_model, ide vs. ahci or such. Anyway I understand you insist on a way to disable it, so I have added the e577927.

I kept the auto-selecting on by default as all migration tests would have to be adjusted as we want this.

On the other hand when directly using qemu monitor I disabled it by default, therefor low-level interface should be clean and high level interface should be benevolent.

Does it make sense like this?

ldoktor commented on virttest/qemu_vm.py:
```diff
@@ -4059,7 +4059,16 @@ def migrate(self, timeout=virt_vm.BaseVM.MIGRATE_TIMEOUT, protocol="tcp",
+                        except qemu_monitor.MonitorNotSupportedError:
+                            # x-multifd-page-count was dropped without
+                            # replacement, ignore this param
+                            continue
```
yep, makes sense...

ldoktor commented:
Changes:

* Added new commit (first) to switch `Monitor` from old to new class type
* Added `logging.warn` when skipping page migration parameter

Not changed:
* https://github.com/avocado-framework/avocado-vt/pull/2022/commits/0a82706dd7b1b7a8b00880e47dbc9b5c5c506f27#r270764774 as I'd prefer to wait for a real usage (negative testing?)

ldoktor commented:
Hello @luckyh, I have added a commit to allow disabling this feature. For the high-level interface (VM.migrate) it is still enabled by default as all migrate tests would have to be adjusted otherwise, but I have disabled it for the low-level interface (qemu_monitor). Let me know whether it's acceptable like this.

Btw I'm using this patch in my daily testing as without it multifd is not usable with qemu-4+...

ldoktor commented:
Still some issues. Anyway does this makes sense to you? Should I polish it, or do you prefer a different solution?

ldoktor commented:
OK, sorry, tired eyes. It should work now, anyway in my opinion it's not needed as people generally want this and for negative testing they can use low-level API to get the supported params/caps...

luckyh commented:
@ldoktor sorry for the delayed response, I know what you are advocating, but the current approach is still not in favor of me. Anyway, it's not a big deal, I'd like to leave the choice to other maintainers since I'm still busy on some other stuff.

Thanks.

ldoktor commented on virttest/qemu_monitor.py:
```diff
@@ -629,6 +631,54 @@ def wait_for_migrate_progress(self, target):
+    def _get_migrate_capability(self, capability,
+                                disable_auto_x_evaluation=True):
+        """
+        Verify the $capability is listed in migrate-capabilitys. If not try
```
sure

balamuruhans replied:
missed to fix the typo.

ldoktor replied:
oups, thanks

ldoktor commented on virttest/qemu_monitor.py:
```diff
@@ -629,6 +631,54 @@ def wait_for_migrate_progress(self, target):
             time.sleep(0.1)
 
+    def _get_migrate_capability(self, capability,
+                                disable_auto_x_evaluation=True):
```
Yes, it is not supported, because it was not supported before.

ldoktor commented:
Thanks @balamuruhans, I don't know what you mean by the query. I am definitely open to suggestion, but I'd like to get things going as qemu-4.0 is already released and multifd migration there does not work on avocado-vt without this patch...

balamuruhans commented on virttest/qemu_monitor.py:
```diff
@@ -2635,20 +2764,23 @@ def set_migrate_parameter(self, parameter, value):
         """
         parameter_info = self.query("migrate-parameters")
+        parameter = self._get_migrate_parameter(parameter,
+                                                disable_auto_x_evaluation=disable_auto_x_evaluation)
```
Something like this,
1. could help to address x-.* attributes in a generic manner
2. could avoid calling self.query() multiple times in multiple places

```python
def x_evaluation(self, attribute, error_on_missing=False):
    if attribute.startswith("x-"):
        attribute2 = attribute[2:]
    else:
        attribute2 = "x-%s" % attribute
    if attribute2 in self._supported_attributes:
        return attribute2
    if error_on_missing:
        raise MonitorNotSupportedError("Attribute %s, nor %s "
                                       "supported." % (attribute,
                                                       attribute2))

def query(self, what, attribute=None, disable_auto_x_evaluation=True):
    info = self.info(what)
    if attribute is None:
        return info
    if self._supported_attributes is None:
                self._supported_attributes = info.keys()
    if(disable_auto_x_evaluation or
       attribute in self._supported_attributes):
        return attribute
    return x_evaluation(attribute)


# make it generic for all the query of parameters and capabilities
def get_migrate_parameter(self, parameter, disable_auto_x_evaluation=True):
    self.query("migrate-parameters", attribute="x-multifd-channels"
               disable_auto_x_evaluation=disable_auto_x_evaluation)

def get_migrate_capability(self, capability,  disable_auto_x_evaluation=True):
    self.query("migrate-capabilities", attribute="x-multifd"
               disable_auto_x_evaluation=disable_auto_x_evaluation)
```

ldoktor replied:
I see. Given the number of usage I simply hardcoded it, but let's give it a try...

ldoktor commented:
@balamuruhans the only change is that I extracted the `x-` evaluation and re-used it everywhere.

ldoktor commented:
@balamuruhans just rebased, no changes. It'd be nice to get this in as qemu-4.0 is almost here and it requires this patch in order to run multifd migration. I've been using this patch since the beginning of 4.0 development and it works well.

balamuruhans commented on virttest/qemu_monitor.py:
```diff
@@ -177,6 +177,49 @@ def wait_for_create_monitor(vm, monitor_name, monitor_params, timeout):
+                             disable_auto_x_evaluation,
+                             error_on_missing=False, feature_type="Feature"):
+    """
+    Attempts to chose supported feature with/without "x-" prefix based
```
%s/chose/choose

ldoktor replied:
OK

balamuruhans commented on virttest/qemu_monitor.py:
```diff
@@ -177,6 +177,49 @@ def wait_for_create_monitor(vm, monitor_name, monitor_params, timeout):
+
+    :param feature: feature that allows x- or non-x- prefix
+    :param supported_features: list of supported features
+    :param error_on_missing: whether to fail when no variant is supported
```
docstring for `disable_auto_x_evaluation` would help

ldoktor replied:
oups, adding

balamuruhans commented on virttest/qemu_monitor.py:
```diff
@@ -2597,23 +2718,42 @@ def set_migrate_capability(self, state, capability):
+                return self.cmd(cmd, args)
+            except QMPCmdError as exc2:
+                logging.debug("Error in set_migrate_capability for %s: %s",
+                              capability, exc)
```
should this be moved to L2723 ?

ldoktor replied:
Nop, at this point both attempts already failed so we log both failures and then postprocess the kind of failures and react accordingly.

balamuruhans commented on virttest/qemu_vm.py:
```diff
@@ -123,6 +123,10 @@ class VM(virt_vm.BaseVM):
+    #: By default translate standard and experimental (prefixed with x-)
+    #: options according to supported parameters/capabilities. To turn this
+    #: off enable this option
+    DISABLE_AUTO_X_MIG_OPTS = False
```
why do we hardcode the option ? should this be a param instead.

ldoktor replied:
Well it was suggested to me in the request to make it tweakable and I do agree it's very rarely to be modified (if ever) so not polluting params makes sense in this case to me. I don't really care, I'd ideally drop this commit and had this always on.

ldoktor commented:
@balamuruhans no changes, just the 2 commit message fixes.

ldoktor commented:
@vivianQizhu @dzhengfy would you please do the second review?

ldoktor commented:
No updates, just rebased as there were conflicts...

ldoktor commented:
Thank you, 2 acks, merging...
